ENH: adds ISMAGS support for directed and multigraph with tests and refactor

dschult opened the pull request:
With this PR, ISMAGS supports directed and multigraph inputs. If inputs do not have the same directed status an exception is raised. But multigraph and graph can be compared. So no restriction there.

Refactored methods `_refine_node_partition`, `_map_nodes`, and `_process_ordered_pair_partitions` to replace the recursive with a (DFS) queue approach.  In addition, the ISMAGS class no longer requires that nodes be sortable. 

Corrected error while handling `candidate_sets` which is a dict-of-set-of-frozensets. For each node, we want to find the minimum length of the frozensets associated with that node. So we should use `min(len(x) for x in candidate_sets[n])` rather than `min(candidate_sets[n], key=len)` which returns the smallest length frozenset. This ensures that the nodes with the fewest candidate mapping-image nodes are considered first, which in theory should return the first isomorphism faster.  (Not a correctness issue -- just a performance hit.)

In switching away from recursive methods, we somehow no longer need the bug-fix for ISMAGS. The bug arose for strange cases where symmetry was getting top/bottom partitions with different numbers of elements. I haven't chased down how that error occurred in the original code, but I suspect we were not respecting the node order or the ordering of partition list by the size of the parts. It doesn't affect the answer, but reduces the number of partitions to consider (the strange cases were aborted once detected).

This PR updates and adds tests especially DiGraph tests from VF, VF2, VF2pp.

Many comment changes, variable name changes and removal or renaming of helper functions. I also removed the manually provided cached properties in favor of just computing those attributes in the `__init__` method (they would never change unless the graphs change which would violate other assumptions).

Names: a "partition" is a list of "parts". Each "part" is a set of nodes.
We should only use the plural "partitions" when we have more than one partition.

I did not change the term "color" to "label" in this PR. I also did not move functions in or out of the class structure (to make a function oriented setup instead of methods).
I did not refactor the methods like `is_isomorphic` to be functions. We can discuss those changes for the whole isomorphism subpackage separately.

I thought about keeping the old code alongside the new, but given the needed correction and the bug-fix shortcut, I don't think we should retain that version. The Sandia folks mentioned that they saw some suspicious parts of the ismags code.

I did not use the Sandia python version of ISMAGS because it was clearly a translation of the original java code, and didn't feel right for NetworkX. It is also a command line tool used for graphs stored in a specific ascii file structure e.g. "XX00YX" to indicate which edges exist between, e.g., 3 nodes and their edge type/color/label.

Are there things I can do to make this easier to review? It's often hard to review the switch from recursive to non-recursive. And we've got 3 such conversions here. Should I split that from the handling of DiGraph changes? It might be worth reading the new ismags.py as its own new file rather than the diff, but I'm not sure.  I will wait a week or two and then review it with fresh eyes if that helps.

Peiffap commented on networkx/algorithms/isomorphism/tests/test_ismags.py:
```diff
@@ -208,48 +215,134 @@ def test_labeled_edges(self):
+    g1b_edges = [
         ["a", "g"],
-        ["a", "h"],
+        ["h", "a"],  #
```
What's the point of these "comments"?

dschult replied:
A comment above states the # symbol indicates an edge with direction swapped between "a" and "b" graphs. I added words to the comments to indicate that.

dschult commented on networkx/algorithms/isomorphism/ismags.py:
```diff
@@ -750,154 +682,206 @@ def _get_permutations_by_length(items):
+                        # Node color compatibility should be taken care of by the
+                        # initial candidate lists made by find_subgraphs
+
+                        # Add gn2_options to the right collection. Since new_candidates
+                        # is a dict-of-sets-of-frozensets of node indices it's
+                        # a bit clunky. We can't do .add without changing the original,
+                        # and + also doesn't work. Could do |, but union is clearer?
+                        new_candidates[sgn2] = new_candidates[sgn2].union(
+                            [frozenset(gn2_options)]
+                        )
```
The `|=` operator changes the existing set and we don't want that. 
The comment implies that "the original" set should not be changed (otherwise we could just use `add(...)`.  And the author deemed `union` as more clear than `|` (also in the comment).  I think this is because `union` works with iterables while `|` only works with sets so "union" is commonly used??.  Note that the object we are currently taking the union of is a list of a frozenset. But that would be easy to switch to a set of a frozenset.

Anyway, we could replace (neither actually fit on a single line due to the indent level)
```
new_candidates[sgn2] = new_candidates[sgn2].union([frozenset(gn2_options)])
```
 with 
```
new_candidates[sgn2] = new_candidates[sgn2] | {frozenset(gn2_options)}
```
I decided the second way invites people to replace it with `|=` more than the first does.  Maybe I should just rewrite the comment so that `|` is clearly not updating the existing set. :)

Peiffap replied:
I misunderstood, my bad! Thanks for clarifying that. I personally like the `a | b` syntax better than `a.union(b)`, but if you think it's more prone to being `|=`ified that's definitely a valid concern.

dschult commented on networkx/algorithms/isomorphism/tests/test_ismags.py:
```diff
@@ -16,89 +20,94 @@ def _matches_to_sets(matches):
+        # However, {5, 9} was considered equivalent to {13, 17}, which is *not*
+        # taken into account in the second refinement, tripping a (former)
+        # assertion failure. Note that this is actually the minimal failing
+        # example.
```
Reworded...

dschult reviewed (commented):
I've tried to address all comments either in the new commits or in the discussions.

dschult commented on networkx/algorithms/isomorphism/tests/test_ismags.py:
```diff
@@ -2,11 +2,15 @@
 import networkx as nx
 from networkx.algorithms import isomorphism as iso
 
+graph_classes = [nx.Graph, nx.DiGraph, nx.MultiGraph, nx.MultiDiGraph]
+
```
Can you slice a fixture?  Also, aren't fixtures really for more complicated tasks? I think of the `mark.parametrize` as easier to find than the definition of the fixture.

dschult commented on networkx/algorithms/isomorphism/ismags.py:
```diff
@@ -975,200 +958,149 @@ def _remove_node(node, nodes, constraints):
+        # refine in all branches of possible refinements (to find symmetries)
+        bots = self._refine_node_partition(graph, new_bot, edge_colors, branch=True)
+        for bot in bots:
+            yield list(new_top), bot
```
Yielding from a generator expression seems like using a new tool for the sake of using it. We don't get any of the advantages of the `yield from` syntax and in this case all the parens get in the way of what is going on. 
I did put in the `list_new_top` var to save recreating the list. :) :)

dschult commented:
I added a (maybe too long) description of the algorithm in the module's doc_string. If its too much into the weeds, we can turn it into comments -- or shift it to a doc_string of maybe the class object.

I also dug into the Ordered-Pair-Paritition (Symmetry) DFS shortcuts. The ISMAGS paper has one fairly simple example and refers to a paper by Kategi, et al. which has more details. It turns out that we were not short-cutting some of the suggested short-cuts. I looked at the DFS nodes we are visiting and pruning and it all makes more sense to me now. Plus I was able to simplify some of the logic for pruning branches and add the missing cases (involving "Matching OPPs"). I hope that code is also more readable. I started getting into moving the OPP related methods to be defied near the process_OPP method, but decided moving methods around is a slippery slope and we should get this in, with a follow-up PR to reorganize the class structure, etc. 

Added tests from the papers. A good one is a graph with two distinct cycles (triangle and square). And I included the Petersen Graph since that is known to have many symmetries for not many nodes.

Lots of misc tweaking here and there too. Hard to stop myself sometimes. But I think its all for the good. I will go through one more time later today for left-behind or commented out lines that should be removed. :)

rossbar commented on networkx/algorithms/isomorphism/ismags.py:
```diff
@@ -89,12 +89,158 @@
+  rejected due to the constraints imposed from symmetry, we prune that subtree and
+  consider a new graph candidate node for that subgraph node. When no more graph
+  candidates remain we backtrack to the previous node in the mapping and consider a
+  new graph candidate for that node. Is we ever to a depth where all subgraph nodes
```
```suggestion
  new graph candidate for that node. If we ever get to a depth where all subgraph nodes
```

rossbar commented on networkx/algorithms/isomorphism/ismags.py:
```diff
@@ -89,12 +89,158 @@
+  restrictions. As described in the ISMAGS paper, [1]_, we store each set of structural
+  restrictions separately as a set of possible candidate nodes rather than computing
+  the intersection of that set with the known graph candidates for the subgraph node.
+  We delay taking the intersection until that node is selected to be in the mapping.
+  While choosing the node with fewest candidates, we avoid computing the intersection
+  by using the size of the minimal set to be intersected rather than the size of the
+  intersection. This may make the node ordering slightly worse via a savings of
+  many intersections most of which are not ever needed.
```
This provides very nice background for some of the comments in the code/from last round of review!

rossbar commented on networkx/algorithms/isomorphism/ismags.py:
```diff
@@ -89,12 +89,158 @@
+  intersection. This may make the node ordering slightly worse via a savings of
+  many intersections most of which are not ever needed.
+
+- With a method that yields subgraph_isomorphisms, we can construct functions like
```
Related to the comment above re: the bulleted list formatting, perhaps move this one above the big explanation!

rossbar commented on networkx/algorithms/isomorphism/ismags.py:
```diff
@@ -289,155 +469,115 @@ def __init__(self, graph, subgraph, node_match=None, edge_match=None, cache=None
+
+    def create_aligned_partitions(self, thing_matcher, sg_things, g_things):
+        """Partitions of `things` (nodes or edges) from subgraph and graph
+        based on function `match`.
```
```suggestion
        based on function `thing_matcher`.
```

rossbar commented on networkx/algorithms/isomorphism/ismags.py:
```diff
@@ -975,200 +1053,205 @@ def _remove_node(node, nodes, constraints):
+            if len(top) > 1 or len(bot) > 1:
+                # ignore parts with > 1 element when they are equal
+                # These are called matching partitions in Katebi 2012.
+                # Symmetries in matching partitions are built by looking only
```
```suggestion
                # Symmetries in matching partitions are built by considering only
```

rossbar commented on networkx/algorithms/isomorphism/ismags.py:
```diff
@@ -975,200 +1053,205 @@ def _remove_node(node, nodes, constraints):
+                    # else load the OPP onto queue for further exploration
+                    # Note that we check for Orbit pruning later because orbits may
+                    # be updated while OPP is sitting on the queue.
+                    # Note: that we check for non-isomorphic OPPs in `_refine_opp`.
```
```suggestion
                    # Note that we check for non-isomorphic OPPs in `_refine_opp`.
```

rossbar reviewed (commented):
Thanks for the updates @dschult , they look great to me!

I'm personally in favor of the giant explainer in the notes section - reading through it helped developed a more concrete idea of some of the key concepts like OPPs, cosets, etc. and how they relate to symmetries and DFS constraints. If we wanted to move the text elsewhere (e.g. the ISMAGS module docs might be a relevant place) that's fine too, but I'm +1 on making sure this text is available somewhere (and would vote to push off deciding where exactly to a followup PR).

I left a bunch of minor suggestions related to grammar/formatting which are generally inconsequential.

Re: the big picture, if you think this is in a state where it's ready to go in (i.e. there's no major changes you think need to be added - we can always continue tweaking) then I'd be +1. For this round I only focused on the changes in 65b133c39, but would happy to go over the entire PR again if you think it's ready!

Peiffap commented:
I'm not sure I'll be able to go over this in depth until next week-end (Oct. 11-12) :(. I haven't forgotten about this PR though.
When is 3.6 planned to be released?

rossbar commented:
> When is 3.6 planned to be released?

There is no exact date, but NX typically does a release in October following the Python version bump.
